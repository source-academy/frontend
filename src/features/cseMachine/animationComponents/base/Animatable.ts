import { NodeConfig } from 'konva/lib/Node';
import { Easings } from 'konva/lib/Tween';

import { Visible } from '../../components/Visible';

export type AnimationConfig = {
  durationMultiplier?: number;
  delayMultiplier?: number;
  easing?: typeof Easings.Linear;
};

export abstract class Animatable extends Visible {
  static key = -1;
  /** Plays the animation, and resolves after the animation is complete */
  abstract animate(): Promise<void>;
  /** Properly dispose of the current animation and ensures that subsequent calls to animate cannot be made */
  abstract destroy(): void;
}

export abstract class AnimatableTo<KonvaConfig extends NodeConfig> extends Visible {
  /** Animates the node to the specified values, and resolves after the animation is complete */
  abstract animateTo(to: Partial<KonvaConfig>, animationConfig?: AnimationConfig): Promise<void>;
  /** Properly dispose of the current animation and ensures that subsequent calls to animate cannot be made */
  abstract destroy(): void;
}
