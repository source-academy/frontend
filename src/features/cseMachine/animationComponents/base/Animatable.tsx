import { NodeConfig } from 'konva/lib/Node';

import { Visible } from '../../components/Visible';

export type AnimationConfig = {
  /** Duration of the animation as a factor of `CSEAnimation.defaultDuration`.
   * For example, if the default duration is 300ms and the duration value is set to `0.5`,
   * the animation would be 150ms long. If unspecified, the duration value will default to `1`.
   */
  duration?: number;

  /** Delay of the animation as a factor of `CSEAnimation.defaultDuration`.
   * For example, if the default duration is 300ms and the delay value is set to `0.5`,
   * the animation would be delayed by 150ms before it starts. If unspecified,
   * the delay value will default to `0`.
   */
  delay?: number;

  /** Easing of the animation. The easing function can be chosen from `Konva.Easings`,
   * or a custom function can also be provided. If unspecified, the defult easing function
   * will be `CSEAnimation.defaultEasing`.
   */
  easing?: (time: number, startValue: number, valueRange: number, duration: number) => number;
};

/** Class that represents a general animatable component, with the key method being `animate` */
export abstract class Animatable extends Visible {
  /** The key value to be used for all `KonvaNodeComponent` to prevent them from reusing
   * previously destroyed konva components. This value is decremented after it is used, to prevent
   * conflicts with the incrementing `Layout.key` value used inside the general layout of
   * other none-animation component. */
  static key = -1;

  /** Starts the animation. `animate` should not be called if the animation is still running.
   * @return a void promise that resolves when the animation is complete */
  abstract animate(animationConfig?: AnimationConfig): Promise<void>;

  /** Properly dispose of the current animation and ensures that subsequent calls to animate cannot be made */
  abstract destroy(): void;
}

/** Class that represents an animatable component, which allows for animations towards new values
 * defined inside the key method `animateTo`
 */
export abstract class AnimatableTo<KonvaConfig extends NodeConfig> extends Visible {
  protected listeners: ((props: Partial<KonvaConfig>) => void)[] = [];

  /** Adds a listener that is called when the animation is running */
  addListener(listener: (props: Partial<KonvaConfig>) => void) {
    this.listeners.push(listener);
  }

  /** Removes the specified listener so it is no longer called when the animation is running */
  removeListener(listener: (props: Partial<KonvaConfig>) => void) {
    const i = this.listeners.indexOf(listener);
    if (i > -1) this.listeners.splice(i, 1);
  }

  /** Animates the node to the specified values, and update any listeners attached to the animation
   * component in every frame of the animation. If multiple calls are made to `animateTo` in the same
   * frame, all animations will correctly run and start at the same time. If there are conflicts in
   * the property names in different `animateTo` calls, the latest call will be prioritised for
   * that particular property.
   * @param to the target props
   * @param animationConfig an optional animation config to customise duration, delay or easing
   * @return a void promise that resolves when the animation is complete
   */
  abstract animateTo(to: Partial<KonvaConfig>, animationConfig?: AnimationConfig): Promise<void>;

  /** Properly dispose of the current animation and ensures that subsequent calls to animate cannot be made */
  abstract destroy(): void;
}
