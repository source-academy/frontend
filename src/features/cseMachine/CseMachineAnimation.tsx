import { InstrType } from 'js-slang/dist/cse-machine/types';
import { Easings } from 'konva/lib/Tween';

import { Animatable } from './animationComponents/AnimationComponents';
import { BinaryOperationAnimation } from './animationComponents/BinaryOperationAnimation';
import { BlockAnimation } from './animationComponents/BlockAnimation';
import { LiteralAnimation } from './animationComponents/LiteralAnimation';
import { PopAnimation } from './animationComponents/PopAnimation';
import { UnaryOperationAnimation } from './animationComponents/UnaryOperationAnimation';
import { isInstr } from './components/ControlStack';
import CseMachine from './CseMachine';
import { Layout } from './CseMachineLayout';

export class CseAnimation {
  private static animationEnabled = false;
  static readonly animationComponents: Animatable[] = [];
  static readonly defaultDuration = 0.3;
  static readonly defaultEasing = Easings.StrongEaseInOut;

  static enableAnimations(): void {
    CseAnimation.animationEnabled = true;
  }

  static disableAnimations(): void {
    CseAnimation.animationEnabled = false;
  }

  private static clearAnimationComponents(): void {
    CseAnimation.animationComponents.length = 0;
  }

  static updateAnimation() {
    CseAnimation.animationComponents.forEach(a => a.destroy());
    CseAnimation.clearAnimationComponents();

    if (!Layout.previousControlComponent) return;
    const lastControlItem = Layout.previousControlComponent.control.peek();
    const lastControlComponent = Layout.previousControlComponent.stackItemComponents.at(-1);
    if (
      !CseAnimation.animationEnabled ||
      !lastControlItem ||
      !lastControlComponent ||
      !CseMachine.getControlStash() // TODO: handle cases where there are environment animations
    ) {
      return;
    }
    let animation: Animatable | undefined;
    if (!isInstr(lastControlItem)) {
      switch (lastControlItem.type) {
        case 'Literal':
          animation = new LiteralAnimation(
            lastControlComponent,
            Layout.stashComponent.stashItemComponents.at(-1)!
          );
          break;
        case 'Program':
        case 'ExpressionStatement':
        case 'VariableDeclaration':
          const currentControlSize = Layout.controlComponent.control.size();
          const previousControlSize = Layout.previousControlComponent.control.size();
          const numOfItems = currentControlSize - previousControlSize + 1;
          if (numOfItems <= 0) break;
          const targetItems = Array.from({ length: numOfItems }, (_, i) => {
            return Layout.controlComponent.stackItemComponents[previousControlSize + i - 1];
          });
          animation = new BlockAnimation(lastControlComponent, targetItems);
          break;
      }
    } else {
      switch (lastControlItem.instrType) {
        case InstrType.RESET:
        case InstrType.WHILE:
        case InstrType.FOR:
        case InstrType.ASSIGNMENT:
          break;
        case InstrType.UNARY_OP:
          animation = new UnaryOperationAnimation(
            lastControlComponent,
            Layout.previousStashComponent.stashItemComponents.at(-1)!,
            Layout.stashComponent.stashItemComponents.at(-1)!
          );
          break;
        case InstrType.BINARY_OP:
          animation = new BinaryOperationAnimation(
            lastControlComponent,
            Layout.previousStashComponent.stashItemComponents.at(-2)!,
            Layout.previousStashComponent.stashItemComponents.at(-1)!,
            Layout.stashComponent.stashItemComponents.at(-1)!
          );
          break;
        case InstrType.POP:
          animation = new PopAnimation(
            lastControlComponent,
            Layout.previousStashComponent.stashItemComponents.at(-1)!
          );
          break;
        case InstrType.APPLICATION:
        case InstrType.BRANCH:
        case InstrType.ENVIRONMENT:
        case InstrType.ARRAY_LITERAL:
        case InstrType.ARRAY_ACCESS:
        case InstrType.ARRAY_ASSIGNMENT:
        case InstrType.ARRAY_LENGTH:
        case InstrType.CONTINUE_MARKER:
        case InstrType.BREAK:
        case InstrType.BREAK_MARKER:
        case InstrType.MARKER:
      }
    }
    if (animation) CseAnimation.animationComponents.push(animation);
  }

  static playAnimation(): void {
    if (!CseAnimation.animationEnabled) {
      CseAnimation.disableAnimations();
      return;
    }
    CseAnimation.disableAnimations();
    for (const animationComponent of this.animationComponents) {
      animationComponent.animate();
    }
  }
}
