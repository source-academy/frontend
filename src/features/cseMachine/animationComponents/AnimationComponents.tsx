import { NodeConfig } from 'konva/lib/Node';
import { RectConfig } from 'konva/lib/shapes/Rect';
import { TextConfig } from 'konva/lib/shapes/Text';
import { Easings, Tween } from 'konva/lib/Tween';
import React from 'react';
import { Group, Rect, Text } from 'react-konva';

import { Visible } from '../components/Visible';
import { CseAnimation } from '../CseMachineAnimation';
import { ControlStashConfig } from '../CseMachineControlStash';
import { currentItemSAColor } from '../CseMachineUtils';

/** Type that extends the NodeConfig type from Konva, making the x, y, width & height values required */
type StrictNodeConfig = Omit<NodeConfig, 'x' | 'y' | 'width' | 'height'> & {
  x: number;
  y: number;
  width: number;
  height: number;
};

type AnimationConfig = {
  durationMultiplier?: number;
  delayMultiplier?: number;
  easing?: typeof Easings.Linear;
};

export abstract class Animatable extends Visible {
  static key = -1;
  /** Plays the animation, and resolves after the animation is complete */
  abstract animate(): Promise<void>;
  /** Properly dispose of the current animation and ensures that subsequent calls to animate cannot be made */
  abstract destroy(): void;
}

abstract class AnimationComponent extends Animatable {
  private isDestroyed = false;
  private tween?: Tween;
  private resolve?: (value: void | PromiseLike<void>) => void;

  constructor(
    protected from: StrictNodeConfig,
    protected to?: NodeConfig,
    protected animationConfig?: AnimationConfig
  ) {
    super();
    this._x = from.x;
    this._y = from.y;
    this._width = from.width;
    this._height = from.height;
  }

  private async delay(duration: number) {
    return new Promise(resolve => {
      setTimeout(resolve, duration);
    });
  }

  setDestination(to: NodeConfig, animationConfig?: AnimationConfig) {
    this.to = to;
    if (animationConfig) this.animationConfig = animationConfig;
  }

  async animate(): Promise<void> {
    return new Promise(async (resolve, reject) => {
      if (this.isDestroyed) {
        resolve();
        return;
      }
      if (!this.ref.current || !this.to) {
        const errorMsg = !this.ref.current
          ? 'Current node reference is null, unable to play animation!'
          : 'No destination value specified!';
        reject('Animation error: ' + errorMsg);
        return;
      }
      this.resolve = resolve;
      if (this.animationConfig?.delayMultiplier) {
        await this.delay(this.animationConfig.delayMultiplier * CseAnimation.defaultDuration);
      }
      if (this.isDestroyed) {
        return;
      }
      if (this.tween) {
        this.tween.pause();
        this.tween.destroy();
      }
      this.tween = new Tween({
        node: this.ref.current,
        ...this.to,
        duration: CseAnimation.defaultDuration * (this.animationConfig?.durationMultiplier ?? 1),
        easing: this.animationConfig?.easing ?? CseAnimation.defaultEasing,
        onFinish: () => {
          if (this.to) {
            if (this.to.x) this._x = this.to.x;
            if (this.to.y) this._y = this.to.y;
            if (this.to.width) this._width = this.to.width;
            if (this.to.height) this._height = this.to.height;
          }
          this.resolve = undefined;
          this.tween?.destroy();
          this.tween = undefined;
          resolve();
        }
      });
      this.tween.play();
    });
  }

  destroy() {
    this.isDestroyed = true;
    if (this.tween) {
      this.tween.finish();
    } else {
      this.resolve?.();
      this.resolve = undefined;
    }
  }
}

export class AnimatedTextComponent extends AnimationComponent {
  readonly text?: string;

  constructor(
    from: StrictNodeConfig,
    to?: NodeConfig,
    private textProps?: TextConfig,
    animationConfig?: AnimationConfig
  ) {
    super(from, to, animationConfig);
    if (this.textProps?.text === undefined) {
      console.warn('AnimatedTextComponent has no text defined inside textProps.');
    }
    this.text = this.textProps?.text;
  }

  draw(): React.ReactNode {
    return (
      <Text
        ref={this.ref}
        {...(this.textProps ?? {})}
        {...this.from}
        text={this.text}
        width={this.width()}
        height={this.height()}
      />
    );
  }
}

export class AnimatedRectComponent extends AnimationComponent {
  constructor(
    from: StrictNodeConfig,
    to?: NodeConfig,
    private rectProps?: RectConfig,
    animationConfig?: AnimationConfig
  ) {
    super(from, to, animationConfig);
  }

  draw(): React.ReactNode {
    return (
      <Rect
        ref={this.ref}
        {...(this.rectProps ?? {})}
        {...this.from}
        width={this.width()}
        height={this.height()}
      />
    );
  }
}

export class AnimatedTextboxComponent extends Animatable {
  private rectComponent: AnimatedRectComponent;
  private textComponent: AnimatedTextComponent;

  constructor(
    from: StrictNodeConfig,
    to?: NodeConfig,
    text?: string,
    animationConfig?: AnimationConfig
  ) {
    super();
    const rectProps = {
      stroke: currentItemSAColor(false),
      cornerRadius: Number(ControlStashConfig.ControlItemCornerRadius)
    };
    const textProps = {
      text,
      fill: ControlStashConfig.SA_WHITE.toString(),
      padding: Number(ControlStashConfig.ControlItemTextPadding),
      fontFamily: ControlStashConfig.FontFamily.toString(),
      fontSize: Number(ControlStashConfig.FontSize),
      fontStyle: ControlStashConfig.FontStyle.toString(),
      fontVariant: ControlStashConfig.FontVariant.toString()
    };
    this.rectComponent = new AnimatedRectComponent(from, to, rectProps, animationConfig);
    this.textComponent = new AnimatedTextComponent(from, to, textProps, animationConfig);
  }

  draw(): React.ReactNode {
    return (
      <Group key={Animatable.key--} ref={this.ref}>
        {this.rectComponent.draw()}
        {this.textComponent.draw()}
      </Group>
    );
  }

  setDestination(to: NodeConfig, animationConfig?: AnimationConfig) {
    this.rectComponent.setDestination(to, animationConfig);
    this.textComponent.setDestination(to, animationConfig);
  }

  async animate() {
    await Promise.all([this.rectComponent.animate(), this.textComponent.animate()]);
  }

  destroy() {
    this.rectComponent.destroy();
    this.textComponent.destroy();
  }
}
