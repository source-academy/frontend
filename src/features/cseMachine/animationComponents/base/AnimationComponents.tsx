import Konva from 'konva';
import { AnimationFn } from 'konva/lib/types';
import React from 'react';
import { Arrow, KonvaNodeComponent, Path, Rect, Text } from 'react-konva';

import { CseAnimation } from '../../CseMachineAnimation';
import { Config } from '../../CseMachineConfig';
import { ControlStashConfig } from '../../CseMachineControlStashConfig';
import { defaultStrokeColor, defaultTextColor } from '../../CseMachineUtils';
import { Animatable, AnimatableTo, AnimationConfig } from './Animatable';
import { lerp } from './AnimationUtils';

type AnimationData<KonvaConfig extends Konva.NodeConfig> = {
  startTime: number;
  endTime: number;
  from: Readonly<Partial<KonvaConfig>>;
  current: Partial<KonvaConfig>;
  to: Readonly<Partial<KonvaConfig>>;
  easing: NonNullable<AnimationConfig['easing']>;
  resolve: (value: void | PromiseLike<void>) => void;
};

abstract class BaseAnimationComponent<
  KonvaConfig extends Konva.NodeConfig
> extends AnimatableTo<KonvaConfig> {
  private isDestroyed = false;
  private animationData: AnimationData<KonvaConfig>[] = [];
  private animation: Konva.Animation;

  private animationFn: AnimationFn = frame => {
    if (!frame || this.animationData.length === 0) return false;
    if (!this.ref.current) {
      this.animationData.forEach(data => data.resolve());
      this.animation.stop();
      return;
    }

    let animationComplete = true;
    const attrs: Partial<KonvaConfig> = {};
    const resolveList: ((value: void | PromiseLike<void>) => void)[] = [];

    let i = 0;
    while (i < this.animationData.length) {
      const data = this.animationData[i];
      if (frame.time <= data.startTime) {
        animationComplete = false;
        i++;
        continue;
      }
      // Calculate animation progress from 0 to 1
      const delta = Math.min((frame.time - data.startTime) / (data.endTime - data.startTime), 1);
      // Interpolate each attribute between the starting and ending values
      for (const attr in data.current) {
        const value = lerp(delta, attr, data.from[attr], data.to[attr], data.easing);
        data.current[attr] = value;
        if (attr === 'x') this._x = value;
        if (attr === 'y') this._y = value;
        if (attr === 'width') this._width = value;
        if (attr === 'height') this._height = value;
      }
      // Add the new attributes and values into the main attrs object
      Object.assign(attrs, data.current);
      // Resolve the animation's promise later if the animation is done, and also
      // remove the animation data from the list
      if (delta === 1) {
        resolveList.push(data.resolve);
        this.animationData.splice(i, 1);
      } else {
        animationComplete = false;
        i++;
      }
    }
    if (Object.keys(attrs).length > 0) {
      // Set all the attributes in one go to improve performance
      this.ref.current.setAttrs(attrs);
      this.listeners.forEach(f => f({ ...attrs }));
    }
    if (animationComplete) this.animation.stop();
    resolveList.forEach(r => r());
    return;
  };

  constructor(protected props: KonvaConfig) {
    super();
    if (!CseAnimation.getLayer()) {
      // If this occurs, it would most likely mean that this animation component is created
      // before the animation layer is even drawn, as the layer ref current value would be null.
      throw new Error('Missing animation layer! Unable to create animation component!');
    }
    this.animation = new Konva.Animation(this.animationFn, CseAnimation.getLayer());
    if (props.x) this._x = props.x;
    if (props.y) this._y = props.y;
    if (props.width) this._width = props.width;
    if (props.height) this._height = props.height;
  }

  animateTo(to: Partial<KonvaConfig>, animationConfig?: AnimationConfig): Promise<void> {
    return new Promise(resolve => {
      // Note: this.ref.current being undefined could also result from not calling the draw
      // function of this animation component.
      // TODO: find a better way to detect the animation component not being drawn
      if (this.isDestroyed || Object.keys(to).length === 0 || !this.ref.current) {
        resolve();
        return;
      }
      const node: Konva.Node = this.ref.current;
      // Get current node values first
      const from: Partial<KonvaConfig> = {};
      for (const attr in to) {
        from[attr] = node.getAttr(attr);
      }
      // Calculate timings based on values given in animationConfig
      const startTime =
        this.animation.frame.time + (animationConfig?.delay ?? 0) * CseAnimation.defaultDuration;
      const endTime = startTime + (animationConfig?.duration ?? 1) * CseAnimation.defaultDuration;
      const easing = animationConfig?.easing ?? CseAnimation.defaultEasing;
      // Add animation data
      const data = { startTime, endTime, from, current: { ...from }, to, easing, resolve };
      this.animationData.push(data);
      // Play animation
      if (!this.animation.isRunning()) this.animation.start();
    });
  }

  destroy() {
    this.isDestroyed = true;
    if (this.animation.isRunning()) {
      this.animation.stop();
    }
    this.ref.current?.hide();
    this.animationData.forEach(data => data.resolve());
    this.animationData.length = 0;
    this.listeners.length = 0;
  }
}

export class AnimationComponent<
  KonvaNode extends Konva.Node,
  KonvaConfig extends Konva.NodeConfig
> extends BaseAnimationComponent<KonvaConfig> {
  constructor(
    /**
     * The `KonvaNodeComponent` that we want to build. Examples: `Text`, `Rect`, `Path`, etc.
     * Note that these components are imported from the 'react-konva' library, not the
     * standard 'konva' library.
     */
    // Note that the React nodes from `react-konva` (such as `Text`, `Rect`, etc.) are not
    // actually types or classes, but are actually variables.
    private type: KonvaNodeComponent<KonvaNode, KonvaConfig>,
    /** The props we want our konva node to have initially. It should match the correct
     *  subtype of `NodeConfig` that the konva node requires. */
    props: KonvaConfig
  ) {
    super(props);
  }

  draw(): React.ReactNode {
    const ReactKonvaNode = this.type;
    return (
      <ReactKonvaNode
        ref={this.ref}
        key={Animatable.key--}
        {...this.props}
        listening={false}
        perfectDrawEnabled={false}
      />
    );
  }
}

export class AnimatedTextComponent extends AnimationComponent<Konva.Text, Konva.TextConfig> {
  constructor(props: Konva.TextConfig & Required<Pick<Konva.TextConfig, 'text'>>) {
    const defaultProps = {
      fill: defaultTextColor(),
      fontFamily: ControlStashConfig.FontFamily,
      fontSize: ControlStashConfig.FontSize,
      fontStyle: ControlStashConfig.FontStyle,
      fontVariant: ControlStashConfig.FontVariant
    };
    super(Text, { ...defaultProps, ...props, width: undefined });
  }

  animateTo(to: Partial<Konva.TextConfig>, animationConfig?: AnimationConfig) {
    return super.animateTo({ ...to, width: undefined }, animationConfig);
  }
}

export class AnimatedRectComponent extends AnimationComponent<Konva.Rect, Konva.RectConfig> {
  constructor(props: Konva.RectConfig) {
    const defaultProps = {
      stroke: defaultStrokeColor(),
      cornerRadius: ControlStashConfig.ControlItemCornerRadius
    };
    super(Rect, { ...defaultProps, ...props });
  }
}

export class AnimatedPathComponent extends AnimationComponent<Konva.Path, Konva.PathConfig> {
  constructor(props: Konva.PathConfig & Required<Pick<Konva.PathConfig, 'data'>>) {
    const defaultProps = {
      stroke: defaultStrokeColor(),
      strokeWidth: Config.ArrowStrokeWidth
    };
    super(Path, { ...defaultProps, ...props });
  }
}

export class AnimatedArrowComponent extends AnimationComponent<Konva.Arrow, Konva.ArrowConfig> {
  constructor(props: Konva.ArrowConfig) {
    const defaultProps = {
      fill: defaultStrokeColor(),
      strokeEnabled: false,
      pointerWidth: Config.ArrowHeadSize
    };
    super(Arrow, { ...defaultProps, ...props });
  }
}
