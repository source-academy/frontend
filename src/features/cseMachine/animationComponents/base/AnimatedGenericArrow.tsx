import { ArrowConfig } from 'konva/lib/shapes/Arrow';
import { PathConfig } from 'konva/lib/shapes/Path';
import { Group } from 'react-konva';
import { SharedProperties } from 'src/commons/utils/TypeHelper';

import { GenericArrow } from '../../components/arrows/GenericArrow';
import { Visible } from '../../components/Visible';
import { defaultStrokeColor, fadedStrokeColor } from '../../CseMachineUtils';
import { Animatable, AnimatableTo, AnimationConfig } from './Animatable';
import { AnimatedArrowComponent, AnimatedPathComponent } from './AnimationComponents';

type PathArrowSharedConfig = Omit<SharedProperties<PathConfig, ArrowConfig>, 'width' | 'height'>;

export class AnimatedGenericArrow<
  Source extends Visible,
  Target extends Visible
> extends AnimatableTo<PathArrowSharedConfig> {
  private pathComponent: AnimatedPathComponent;
  private arrowComponent: AnimatedArrowComponent;

  private onPropsChange = (props: PathConfig) => {
    if (props.x) this._x = this.arrow.x() + props.x;
    if (props.y) this._y = this.arrow.y() + props.y;
  };

  constructor(
    private arrow: GenericArrow<Source, Target>,
    props?: PathArrowSharedConfig
  ) {
    super();
    this._x = arrow.x();
    this._y = arrow.y();
    this._width = arrow.width();
    this._height = arrow.height();
    this.pathComponent = new AnimatedPathComponent({
      stroke: arrow.faded ? fadedStrokeColor() : defaultStrokeColor(),
      data: arrow.path(),
      ...props
    });
    this.pathComponent.addListener(this.onPropsChange);
    this.arrowComponent = new AnimatedArrowComponent({
      fill: arrow.faded ? fadedStrokeColor() : defaultStrokeColor(),
      points: arrow.points.slice(arrow.points.length - 4),
      ...props
    });
  }

  draw(): React.ReactNode {
    return (
      <Group ref={this.ref} key={Animatable.key--}>
        {this.pathComponent.draw()}
        {this.arrowComponent.draw()}
      </Group>
    );
  }

  animatePathTo(to: Partial<PathConfig>, animationConfig?: AnimationConfig) {
    return this.pathComponent.animateTo(to, animationConfig);
  }

  animateArrowTo(to: Partial<ArrowConfig>, animationConfig?: AnimationConfig) {
    return this.arrowComponent.animateTo(to, animationConfig);
  }

  async animateTo(to: Partial<PathArrowSharedConfig>, animationConfig?: AnimationConfig) {
    await Promise.all([
      this.animatePathTo(to, animationConfig),
      this.animateArrowTo(to, animationConfig)
    ]);
  }

  destroy() {
    this.pathComponent.removeListener(this.onPropsChange);
    this.pathComponent.destroy();
    this.arrowComponent.destroy();
  }
}
