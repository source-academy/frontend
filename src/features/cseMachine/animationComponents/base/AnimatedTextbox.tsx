import { RectConfig } from 'konva/lib/shapes/Rect';
import { TextConfig } from 'konva/lib/shapes/Text';
import { Group } from 'react-konva';
import { SharedProperties } from 'src/commons/utils/TypeHelper';

import { ControlStashConfig } from '../../CseMachineControlStashConfig';
import { Animatable, AnimatableTo, AnimationConfig } from './Animatable';
import { AnimatedRectComponent, AnimatedTextComponent } from './AnimationComponents';

type TextRectSharedConfig = SharedProperties<TextConfig, RectConfig>;

export class AnimatedTextbox extends AnimatableTo<TextRectSharedConfig> {
  private rectComponent: AnimatedRectComponent;
  private textComponent: AnimatedTextComponent;

  // Update the current component's dimensions based on the inner rect component's dimensions
  private onPropsChange = (props: RectConfig) => {
    if (props.x) this._x = props.x;
    if (props.y) this._y = props.y;
    if (props.width) this._width = props.width;
    if (props.height) this._height = props.height;
  };

  constructor(
    text: string,
    sharedProps: TextRectSharedConfig,
    additionalProps?: { rectProps?: RectConfig; textProps?: TextConfig }
  ) {
    super();
    const rectProps = { ...sharedProps, ...additionalProps?.rectProps };
    this.onPropsChange(rectProps);
    this.rectComponent = new AnimatedRectComponent(rectProps);
    this.rectComponent.addListener(this.onPropsChange);
    const textProps = {
      padding: ControlStashConfig.ControlItemTextPadding,
      ...sharedProps,
      ...additionalProps?.textProps,
      text
    };
    this.textComponent = new AnimatedTextComponent(textProps);
  }

  draw(): React.ReactNode {
    return (
      <Group ref={this.ref} key={Animatable.key--}>
        {this.rectComponent.draw()}
        {this.textComponent.draw()}
      </Group>
    );
  }

  animateRectTo(to: Partial<RectConfig>, animationConfig?: AnimationConfig) {
    return this.rectComponent.animateTo(to, animationConfig);
  }

  animateTextTo(to: Partial<TextConfig>, animationConfig?: AnimationConfig) {
    return this.textComponent.animateTo(to, animationConfig);
  }

  async animateTo(to: Partial<TextRectSharedConfig>, animationConfig?: AnimationConfig) {
    await Promise.all([
      this.animateRectTo(to, animationConfig),
      this.animateTextTo(to, animationConfig)
    ]);
  }

  destroy() {
    this.rectComponent.removeListener(this.onPropsChange);
    this.rectComponent.destroy();
    this.textComponent.destroy();
  }
}
