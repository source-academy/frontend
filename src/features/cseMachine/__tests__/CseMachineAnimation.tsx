import Konva from 'konva';
import * as ReactKonva from 'react-konva';

import { AnimationConfig } from '../animationComponents/base/Animatable';
import { AnimationComponent } from '../animationComponents/base/AnimationComponents';
import { CseAnimation } from '../CseMachineAnimation';

const mockStage = new Konva.Stage({
  container: document.createElement('div'),
  width: 500,
  height: 500
} as Konva.StageConfig);
const mockLayer = new Konva.Layer();
mockStage.add(mockLayer);
// Override layer ref's current property to the mock layer
Object.defineProperty(CseAnimation.layerRef, 'current', { value: mockLayer });

// The points in the animation which the checker should check for.
const deltaIntervals = [0, 0.2, 0.4, 0.5, 0.6, 0.8, 1];

async function checkComponentAnimation<
  KonvaNode extends Konva.Node,
  KonvaConfig extends Konva.NodeConfig
>(
  component: AnimationComponent<KonvaNode, KonvaConfig>,
  initialProps: KonvaConfig,
  targetProps: Partial<KonvaConfig>,
  animationConfig: Required<AnimationConfig>
) {
  const duration = animationConfig.duration * CseAnimation.defaultDuration;
  const delay = animationConfig.delay * CseAnimation.defaultDuration;
  const easing = animationConfig.easing;
  const node = component.ref.current as KonvaNode;
  const timings = deltaIntervals.map(d => d * duration + delay);

  let i = 0;
  const startTime = performance.now();
  const checker = () => {
    const elapsed = performance.now() - startTime - delay;
    if (timings[i] - elapsed < 50 / 3 || elapsed > timings[i]) {
      for (const [attr, target] of Object.entries(targetProps)) {
        const initial = initialProps[attr];
        if (initial === undefined) {
          // Check that animation is skipped for props with no initial values
          expect(node.getAttr(attr)).toEqual(target);
          continue;
        }

        const actual = node.getAttr(attr) as number;
        const expected = easing(elapsed, initial, target - initial, duration);

        // Tolerance is the max difference between actual and expected.
        // Example: initial value = 100, target value = 200, duration = 300ms => tolerance = 2
        // Longer duration => lower tolerance. Larger difference in values => higher tolerance
        const tolerance = (Math.abs(target - initial) / duration) * 6;

        // See: https://github.com/jestjs/jest/blob/main/packages/expect/src/matchers.ts#L144
        // Below calculation for precision is the inverse of the calculation in `toBeCloseTo`
        expect(actual).toBeCloseTo(expected, -Math.log10(2 * tolerance));
      }
      if (i < timings.length - 1) {
        i++;
        requestAnimationFrame(checker);
      }
    } else {
      requestAnimationFrame(checker);
    }
  };
  requestAnimationFrame(checker);
}

// Mapping of React Konva components to Konva nodes
const konvaNodeMap = {
  Label: Konva.Label,
  Rect: Konva.Rect,
  Circle: Konva.Circle,
  Ellipse: Konva.Ellipse,
  Wedge: Konva.Wedge,
  Line: Konva.Line,
  Sprite: Konva.Sprite,
  Image: Konva.Image,
  Text: Konva.Text,
  TextPath: Konva.TextPath,
  Star: Konva.Star,
  Ring: Konva.Ring,
  Arc: Konva.Arc,
  Tag: Konva.Tag,
  Path: Konva.Path,
  RegularPolygon: Konva.RegularPolygon,
  Arrow: Konva.Arrow,
  Shape: Konva.Shape
};

async function testSingleAnimation<
  KonvaNode extends Konva.Node,
  KonvaConfig extends Konva.NodeConfig
>(
  component: AnimationComponent<KonvaNode, KonvaConfig>,
  targetProps: Partial<KonvaConfig>,
  animationConfig?: AnimationConfig
) {
  for (const value of Object.values(targetProps)) {
    if (typeof value !== 'number') {
      throw new Error('Checking of non-number animation values is not supported in testing!');
    }
  }

  // React's `RefObject`s don't work properly during testing, so we have to access internal
  // properties of AnimationComponent and manually create the Konva nodes
  const type = (component as any).type as keyof typeof konvaNodeMap;
  const initialProps = (component as any).props as KonvaConfig;
  expect(konvaNodeMap[type]).toBeDefined();
  const node = Reflect.construct(konvaNodeMap[type], [initialProps]) as Konva.Shape;
  Object.defineProperty(component.ref, 'current', { value: node });
  mockLayer.add(node);

  // Peek into node attributes and check that the values match while the animation is running
  const config = {
    duration: animationConfig?.duration ?? 1,
    delay: animationConfig?.delay ?? 0,
    easing: animationConfig?.easing ?? CseAnimation.defaultEasing
  };

  // Run animation and checks in parallel
  await Promise.all([
    component.animateTo(targetProps, animationConfig),
    checkComponentAnimation(component, initialProps, targetProps, config)
  ]);

  // Check at the end that all target props have been fulfilled
  for (const [attr, target] of Object.entries(targetProps)) {
    expect(node.getAttr(attr)).toEqual(target);
  }

  node.destroy();
}

test('AnimationComponent animates correctly with default animation config', async () => {
  const start = { height: 100 };
  const target = { height: 200 };
  const component = new AnimationComponent(ReactKonva.Rect, start);
  expect(component.draw()).toMatchSnapshot();
  await testSingleAnimation(component, target);
});

test('AnimationComponent animates correctly with custom animation config', async () => {
  const start = { x: 0, y: 100, opacity: 0, text: 'test' };
  const target = { x: 400, y: 200, opacity: 1 };
  const component = new AnimationComponent(ReactKonva.Text, start);
  expect(component.draw()).toMatchSnapshot();
  await testSingleAnimation(component, target, {
    duration: 1.5,
    delay: 0.5,
    easing: (t, b, c, d) => b + (t /= d) * t * t * c // EaseInCubic
  });
});
