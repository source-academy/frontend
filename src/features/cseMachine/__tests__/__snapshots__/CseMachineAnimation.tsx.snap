// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimationComponent animates correctly with custom animation config 1`] = `
<Text
  listening={false}
  opacity={0}
  perfectDrawEnabled={false}
  text="test"
  x={0}
  y={100}
/>
`;

exports[`AnimationComponent animates correctly with default animation config 1`] = `
<Rect
  height={100}
  listening={false}
  perfectDrawEnabled={false}
/>
`;
